ui/dashboards/src/utils/panelUtils.ts

```diff
@@ -26,6 +27,130 @@ export function getYForNewRow(group: PanelGroupDefinition) {
+ * reference panel. The new panel is placed:
+ * - To the right of the reference panel if there is space available without
+ *   moving other panels.
+ * - Otherwise, directly below the reference panel. If other panels are below
+ *   this location, they will also shift downward because the grid uses
+ *   vertical-based compacting.
+ *
+ * @param newLayout - Layout for new panel to insert into the grid.
+ * @param referenceLayout - Layout for reference panel used to determine the
+ *   placement of the new panel.
+ * @param itemLayouts - Full grid layout.
+ * @returns - Item layouts modified to insert the new panel.
+ */
+export function insertPanelInLayout(
+  newLayout: UnpositionedPanelGroupItemLayout,
+  referenceLayout: PanelGroupItemLayout,
+  itemLayouts: PanelGroupItemLayout[]
+): PanelGroupItemLayout[] {
+  const MAX_LAYOUT_WIDTH = GRID_LAYOUT_COLS[GRID_LAYOUT_SMALL_BREAKPOINT];
+
+  const referenceBounds = getPanelBounds(referenceLayout);
+
+  // Organize layouts based on vertical relation to the item being inserted
+  // after.
+  const aboveInsertRow: PanelGroupItemLayout[] = [];
+  const insertRow: PanelGroupItemLayout[] = [];
+  const belowInsertRow: PanelGroupItemLayout[] = [];
+  itemLayouts.forEach((itemLayout) => {
+    const itemBounds = getPanelBounds(itemLayout);
+
+    if (itemBounds.y2 <= referenceBounds.y1) {
+      aboveInsertRow.push(itemLayout);
+    } else if (itemBounds.y1 >= referenceBounds.y2) {
+      belowInsertRow.push(itemLayout);
+    } else {
+      insertRow.push(itemLayout);
+    }
+  });
+
+  // Cannot safely assume that the order of item layouts array is strictly
+  // left to right. Sorting the row by horizontal position to more easily find
+  // gaps.
+  insertRow.sort((a, b) => a.x - b.x);
+  const insertAfterIndex = insertRow.findIndex((item) => item.i === referenceLayout.i);
+
+  if (insertAfterIndex === insertRow.length - 1) {
+    // Insert to the right when space is available and the reference is the last
+    // item in the row.
+    if (referenceBounds.x2 + newLayout.w <= MAX_LAYOUT_WIDTH) {
+      return [
+        ...aboveInsertRow,
+        ...insertRow,
+        {
+          ...newLayout,
+          x: referenceBounds.x2,
+          y: referenceBounds.y1,
+        },
+        ...belowInsertRow,
+      ];
+    }
+  } else if (insertAfterIndex >= 0) {
+    const nextItem = insertRow[insertAfterIndex + 1];
+
+    if (nextItem && getPanelBounds(nextItem).x1 - referenceBounds.x2 >= newLayout.w) {
+      // Insert to the right when space is available between the reference and
+      // the next item in the row.
+      insertRow.splice(insertAfterIndex + 1, 0, {
+        ...newLayout,
+        x: referenceBounds.x2,
+        y: referenceBounds.y1,
+      });
+
+      return [...aboveInsertRow, ...insertRow, ...belowInsertRow];
+    }
+  }
+
+  // Insert the new item below the original and shift the items below the
+  // row where the reference is located.
+  return [
```

Comment:
While I'm bringing up edge cases, this is completely unrelated to your PR, but I wonder if we should be setting a min width and min height? Currently, you can resize the panel down to a tiny, unusable size: 
![image](https://user-images.githubusercontent.com/9369625/218797567-fda592fd-1e29-4452-9dcb-a8f7ec62fcd7.png)

Reply:
> One edge case I found is in the attached recording, where there is room to the right, but the panel is placed underneath. I don't think it's important to address though, I was intentionally trying weird layouts where panels weren't quite aligned--when using the dashboard normally, it works as expected.

In that case, it assumes the panel to the right of it is in the same row because it has vertical overlap, so it doesn't see space to the right and falls back to going below. I agree with your assessment that it's probably not worth doing something different there. It would increase the complexity of the algorithm without delivering a significantly better UX.

Reply:
> While I'm bringing up edge cases, this is completely unrelated to your PR, but I wonder if we should be setting a min width and min height? Currently, you can resize the panel down to a tiny, unusable size: ![image](https://user-images.githubusercontent.com/9369625/218797567-fda592fd-1e29-4452-9dcb-a8f7ec62fcd7.png)

Good call. I'll make a shortcut ticket for this, so we can prioritize fixing it.

Reply:
Created https://app.shortcut.com/chronosphere/story/43552/implement-min-width-height-for-panels-in-dashboard?stories_sort_by=priority&stories_group_by=workflow_state_id